R/TealAppDriver.R

```diff
@@ -554,19 +553,50 @@ TealAppDriver <- R6::R6Class( # nolint: object_name.
       filter_panel = character(0)
     ),
     # private methods ----
+    # Helper function to extract wrapper ID from selector and take first match if multiple found
+    extract_wrapper_id = function(selector) {
+      id <- self$get_attr(selector = selector, attribute = "href")
+      sub("^#", "", id[endsWith(id, "-wrapper")])
+    },
+    # Helper function to check if wrapper ID is valid
+    is_valid_wrapper_id = function(wrapper_id) {
+      length(wrapper_id) >= 1 & wrapper_id != "" & !is.na(wrapper_id)
+    },
     set_active_ns = function() {
+      # Although wait_for_idle() is called before set_active_ns(), it only ensures Shiny is not processing.
+      # wait_for_page_stability() is needed here to ensure the DOM/UI is fully rendered and stable
+      # before trying to extract the namespace.
+      private$wait_for_page_stability()
+
       all_inputs <- self$get_values()$input
       active_tab_inputs <- all_inputs[grepl("-active_module_id$", names(all_inputs))]
```

Comment:
We could use [AppDriver$wait_for_value](https://rstudio.github.io/shinytest2/reference/AppDriver.html?q=get_download#method-wait-for-value-) to make sure we have a value and then discard the rest of the added complexity.

The `ignore = list(NULL, "")` argument will allow us to avoid any of the initialization

spoiler... same can be done with javascript code and `AppDriver$wait_for_js` (let's say for  `is_visible()`)... I'm trying it out in widgets.